### PR TITLE
optimize __neg__ routines

### PR DESCRIPTION
**Pull request by smichr**
#### other issues

#13645 is not fixed

#### Brief description of what is fixed or changed

From #17242 I found the test runtimes decreased significantly when `__neg__` terms were rewritten. I am pushing this change alone to confirm this.

While working on this, a recursion error was encountered during the test run. So, serendipitously, that issue also got fixed. Running `as_real_imag` tests has reduced from about 12 seconds to 8 seconds.

#### Release Notes

* core
    * recusion error involving as_real_imag was removed
    * negation of expressions has been optimized
    * `extract_multiplicatively` returns a canonical expression

### CODE REVIEW

**Comment by smichr** [comment]
When I run tests of core in this PR they run in about half the time as the tests run under master. And this is repeatable as I switch back and forth between the two branches. But running in Travis, I don't see an overall dramatic time difference. Could anyone else confirm the difference in runtime for core?

**Comment by oscarbenjamin** [comment]
I get (master):
```
$ pytest sympy/core/ -n4
1668 passed, 68 skipped, 43 xfailed in 62.65 seconds
```
And with this PR:
```
$ pytest sympy/core/ -n4
1668 passed, 68 skipped, 43 xfailed in 40.98 seconds
```
So it's about 1.5x faster. Good work!

**Comment by smichr** [comment]
```python
>>> pre = 'from random import random\nfrom sympy.abc import x, y\nfrom sympy import S,Mul'

>>> timeit('Mul(S.NegativeOne, random()*x)', pre, number=2000)
3.490064883999821
>>> timeit('-(random()*x)', pre, number=2000)
2.5226658450001196

>>> timeit('0-(random()*x+y)', pre, number=2000)  # treat as Mul(-1, Add)
6.331252217999918
>>> timeit('-(random()*x+y)', pre, number=2000)
4.833894029000021
```

**Comment by smichr** [comment]
Not sure where/why the tests are failing. The `test_simplify.py` suite passes under 3.7 locally.

File: sympy/simplify/tests/test_simplify.py [hunk 1]
```diff
@@ -829,11 +829,11 @@ def test_issue_17141():
     assert simplify(acos(-I)**2*acos(I)**2) == \
            log(1 + sqrt(2))**4 + pi**2*log(1 + sqrt(2))**2/2 + pi**4/16
     assert simplify(2**acos(I)**2) == 2**((pi - 2*I*log(1 + sqrt(2)))**2/4)
+    p = 2**acos(I+1)**2
+    assert simplify(p) == p
 
     # However, for a complex number it still happens
     if PY3:
-        raises(RecursionError, lambda: simplify(2**acos(I+1)**2))
         raises(RecursionError, lambda: simplify((2**acos(I+1)**2).rewrite('log')))
```
Comment (oscargus) [inline sympy/simplify/tests/test_simplify.py]: I should really have added these in a separate test with XFAIL. "Fortunately" it wasn't the only error in the test run.

Comment (smichr) [reply sympy/simplify/tests/test_simplify.py]: I wish that both were solved...the `as_real_imag` needs work (see #6679).

**Comment by smichr** [comment]
If tests pass, this is ready to go.

File: sympy/simplify/tests/test_simplify.py (hunk 1 above)
Comment (oscargus) [reply sympy/simplify/tests/test_simplify.py]: If I could figure out (or someone else tell me, more likely) what `atan2(0, complex)` (see #17155) is, it should (probably) be solved. This is where it starts looping.

Comment (smichr) [reply sympy/simplify/tests/test_simplify.py]: If a and b are real then `1/(a +I*b)` cand be rewritten over a real value...why not just rewrite like that?

Comment (oscargus) [reply sympy/simplify/tests/test_simplify.py]: I think that is what causes the recursion error.

Comment (oscargus) [reply sympy/simplify/tests/test_simplify.py]: Well, not the original expression of course.... The problem is just to figure out when it works or not. (I cannot remember the details in this case but I wrote them down in some issue...)

**Comment by oscargus** [comment]
Looks good, thanks! Merging.

**Pull request merged**